### PR TITLE
feature(davide): Implementation of D4PG

### PR DESCRIPTION
**Pull request by davide97l**
## Description
This PR is to implement the algorithm D4PG: https://arxiv.org/abs/1804.08617. It currently supports n-step, experience replay, and categorical distribution.

### CODE REVIEW

**Comment by davide97l** [comment]
Training reward on Pendulum
image

File: dizoo/classic_control/pendulum/entry/pendulum_d4pg_main.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        SampleSerialCollector,
+        InteractionSerialEvaluator,
+        AdvancedReplayBuffer,
+        save_cfg=True
+    )
+
+    # Set up envs for collection and evaluation
+    collector_env_num, evaluator_env_num = cfg.env.collector_env_num, cfg.env.evaluator_env_num
+    collector_env = BaseEnvManager(
+        env_fn=[lambda: PendulumEnv(cfg.env) for _ in range(collector_env_num)], cfg=cfg.env.manager
```
Comment (PaParaZz1) [inline dizoo/classic_control/pendulum/entry/pendulum_d4pg_main.py:30]: Whether you know the difference between lambda expression and partial function when implements multiple env function, especially in the cases where each env has its own unique cfg

File: dizoo/classic_control/pendulum/entry/pendulum_d4pg_main.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                break
+        # Collect data from environments
+        new_data = collector.collect(train_iter=learner.train_iter)
+        replay_buffer.push(new_data, cur_collector_envstep=collector.envstep)
+        # Train
+        for i in range(cfg.policy.learn.update_per_collect):
+            train_data = replay_buffer.sample(learner.policy.get_attribute('batch_size'), learner.train_iter)
+            if train_data is None:
+                break
+            learner.train(train_data, collector.envstep)
```
Comment (PaParaZz1) [inline dizoo/classic_control/pendulum/entry/pendulum_d4pg_main.py:71]: How do you update priority? I guess you maybe omit it. You can refer to the implementation in `serial_pipeline`

File: ding/policy/d4pg.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+        # =============
+        # after update
+        # =============
+        loss_dict['total_loss'] = sum(loss_dict.values())
+        self._forward_learn_cnt += 1
+        self._target_model.update(self._learn_model.state_dict())
+        return {
+            'cur_lr_actor': self._optimizer_actor.defaults['lr'],
+            'cur_lr_critic': self._optimizer_critic.defaults['lr'],
+            # 'q_value': np.array(q_value).mean(),
```
Comment (PaParaZz1) [inline ding/policy/d4pg.py:278]: I think we should record q_value default, maybe you can enable this line.

File: ding/policy/d4pg.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
+from typing import List, Dict, Any, Tuple, Union
```
Comment (PaParaZz1) [inline ding/policy/d4pg.py:1]: You should add unittest and algotest for d4pg, like that in `ding/entry/tests`.

**Review by PaParaZz1 — APPROVED** [review]
This PR can be merged after solving these problems.

File: ding/model/template/qac_dist.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+        Critic Shapes:
+            - obs (:obj:`torch.Tensor`): :math:`(B, N1)`, where B is batch size and N1 is ``obs_shape``
+            - action (:obj:`torch.Tensor`): :math:`(B, N2)`, where B is batch size and N2 is``action_shape``
+            - q_value (:obj:`torch.FloatTensor`): :math:`(B, N2)`, where B is batch size and N2 is ``action_shape``
+            - distribution (:obj:`torch.FloatTensor`): :math:`(B, 1, N3)`, where B is batch size and N3 is ``num_atom``
+
+        Actor Examples:
+            >>> # Regression mode
+            >>> model = QAC(64, 64, 'regression')
```
[the remark this answers is not in the document]
Comment (davide97l) [reply ding/model/template/qac_dist.py:143]: Updated like this
```
            Critic Examples: 
                    >>> # Categorical mode
                    >>> inputs = {'obs': torch.randn(4,N), 'action': torch.randn(4,1)}
                    >>> model = QACDIST(obs_shape=(N, ),action_shape=1,actor_head_type='regression', \
                    ...                 critic_head_type='categorical', n_atoms=51)
                    >>> q_value = model(inputs, mode='compute_critic') # q value
                    >>> assert q_value['q_value'].shape == torch.Size([4, 1])
                    >>> assert q_value['distribution'].shape == torch.Size([4, 1, 51])
```

File: ding/model/template/qac_dist.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+                    actor_head_hidden_size,
+                    action_shape,
+                    actor_head_layer_num,
+                    sigma_type='conditioned',
+                    activation=activation,
+                    norm_type=norm_type
+                )
+            )
+        self.critic_head_type = critic_head_type
+        assert self.critic_head_type in ['categorical']
```
[the remark this answers is not in the document]
Comment (davide97l) [reply ding/model/template/qac_dist.py:88]: done

File: dizoo/classic_control/pendulum/entry/pendulum_d4pg_main.py (hunk 2 above)
Comment (davide97l) [reply dizoo/classic_control/pendulum/entry/pendulum_d4pg_main.py:71]: Thank you for reminding me, I used a copy of the entry of ddpg as baseline and forgot to add it. I added now: `replay_buffer.update(learner.priority_info)` in line 72
